File: windows/Hinge.Core/DeviceRegistry.cs
using System.Collections.Concurrent;

namespace Hinge.Core;

public class DeviceRecord
{
    public Device Device { get; }
    public DateTime LastSeen { get; set; }

    public DeviceRecord(Device device, DateTime lastSeen)
    {
        Device = device;
        LastSeen = lastSeen;
    }
}

public class DeviceRegistry
{
    private readonly ConcurrentDictionary<string, DeviceRecord> _devices = new();
    private readonly object _sync = new();

    public event EventHandler<IReadOnlyList<Device>>? DevicesChanged;

    public IReadOnlyList<Device> GetAllDevices()
    {
        lock (_sync)
        {
            return _devices.Values.Select(r => r.Device).ToList();
        }
    }

    public bool TryGetDevice(string deviceId, out Device? device)
    {
        lock (_sync)
        {
            if (_devices.TryGetValue(deviceId, out var record))
            {
                device = record.Device;
                return true;
            }
        }
        device = null;
        return false;
    }

    public void UpsertDevice(DiscoveryMessage message, string remoteAddress)
    {
        var now = DateTime.UtcNow;
        bool isNewOrChanged = false;
        DeviceRecord? replacedRecord = null;

        lock (_sync)
        {
            // Reinstalling or migrating the Android package can regenerate its
            // persisted device ID. If the old record is already offline but has
            // the same identity and LAN address, collapse it into the new ID
            // instead of showing two rows for one phone.
            if (!_devices.ContainsKey(message.DeviceId))
            {
                foreach (var pair in _devices.ToArray())
                {
                    if (!CanCollapseStaleIdentity(pair.Value.Device, message, remoteAddress))
                    {
                        continue;
                    }

                    if (_devices.TryRemove(pair.Key, out replacedRecord))
                    {
                        isNewOrChanged = true;
                        break;
                    }
                }
            }

            _devices.AddOrUpdate(
                message.DeviceId,
                _ =>
                {
                    isNewOrChanged = true;
                    return new DeviceRecord(new Device
                    {
                        DeviceId = message.DeviceId,
                        Name = DisplayName(message.Name, message.Manufacturer, message.Model),
                        Manufacturer = message.Manufacturer,
                        Model = message.Model,
                        Platform = Enum.TryParse<DevicePlatform>(message.Platform, true, out var p) ? p : DevicePlatform.Unknown,
                        AppVersion = message.Version,
                        ProtocolVersion = message.ProtocolVersion,
                        SessionPort = message.Port > 0 ? message.Port : Constants.SessionTcpPort,
                        DiscoveryPort = ValidDiscoveryPort(message.DiscoveryPort),
                        Capabilities = message.Capabilities,
                        NetworkAddresses = MergeAddresses(
                            replacedRecord?.Device.NetworkAddresses,
                            remoteAddress),
                        ConnectionState = ConnectionState.Discovered,
                        TrustState = replacedRecord?.Device.TrustState ?? TrustState.Untrusted
                    }, now);
                },
                (_, existing) =>
                {
                    existing.LastSeen = now;
                    var dev = existing.Device;
                    int addressIndex = dev.NetworkAddresses.FindIndex(address =>
                        string.Equals(address, remoteAddress, StringComparison.OrdinalIgnoreCase));
                    if (addressIndex < 0)
                    {
                        dev.NetworkAddresses.Add(remoteAddress);
                        isNewOrChanged = true;
                    }
                    if (dev.ConnectionState == ConnectionState.Disconnected)
                    {
                        dev.ConnectionState = ConnectionState.Discovered;
                        isNewOrChanged = true;
                    }
                    var displayName = DisplayName(message.Name, message.Manufacturer, message.Model);
                    if (dev.Name != displayName ||
                        dev.Manufacturer != message.Manufacturer ||
                        dev.Model != message.Model ||
                        dev.SessionPort != (message.Port > 0 ? message.Port : Constants.SessionTcpPort) ||
                        dev.DiscoveryPort != ValidDiscoveryPort(message.DiscoveryPort))
                    {
                        dev.Name = displayName;
                        dev.Manufacturer = message.Manufacturer;
                        dev.Model = message.Model;
                        dev.SessionPort = message.Port > 0 ? message.Port : Constants.SessionTcpPort;
                        dev.DiscoveryPort = ValidDiscoveryPort(message.DiscoveryPort);
                        isNewOrChanged = true;
                    }
                    return existing;
                }
            );

            // A record can become offline after the replacement check above has
            // already run. Reconcile on every packet so a stale row is removed
            // on the next announcement instead of living beside the live row.
            isNewOrChanged |= ReconcileDuplicateRecords(message.DeviceId);
        }

        if (isNewOrChanged)
        {
            NotifyChanged();
        }
    }

    public void PruneOffline(TimeSpan timeout)
    {
        var cutoff = DateTime.UtcNow - timeout;
        bool changed = false;

        lock (_sync)
        {
            foreach (var pair in _devices)
            {
                if (pair.Value.Device.ConnectionState == ConnectionState.Connected)
                {
                    // A live TCP session is authoritative. UDP discovery packets can be
                    // lost briefly without making an established session offline.
                    continue;
                }
                if (pair.Value.LastSeen < cutoff && pair.Value.Device.ConnectionState != ConnectionState.Disconnected)
                {
                    pair.Value.Device.ConnectionState = ConnectionState.Disconnected;
                    changed = true;
                }
            }
        }

        if (changed)
        {
            NotifyChanged();
        }
    }

    public void MarkSessionDisconnected(string deviceId)
    {
        bool changed = false;
        lock (_sync)
        {
            if (string.IsNullOrWhiteSpace(deviceId) ||
                !_devices.TryGetValue(deviceId, out var record) ||
                record.Device.ConnectionState != ConnectionState.Connected)
            {
                return;
            }

            bool stillPresent = DateTime.UtcNow - record.LastSeen <= TimeSpan.FromSeconds(30);
            record.Device.ConnectionState = stillPresent
                ? ConnectionState.Discovered
                : ConnectionState.Disconnected;
            changed = true;
        }

        if (changed)
        {
            NotifyChanged();
        }
    }

    private bool ReconcileDuplicateRecords(string preferredDeviceId)
    {
        if (!_devices.TryGetValue(preferredDeviceId, out var preferredRecord))
        {
            return false;
        }

        var group = _devices
            .ToArray()
            .Where(pair => pair.Key == preferredDeviceId ||
                CanMergeDuplicateDevices(preferredRecord.Device, pair.Value.Device))
            .ToList();
        if (group.Count <= 1)
        {
            return false;
        }

        var survivor = group
            .OrderByDescending(pair => pair.Value.Device.ConnectionState == ConnectionState.Connected)
            .ThenByDescending(pair => string.Equals(pair.Key, preferredDeviceId, StringComparison.OrdinalIgnoreCase))
            .ThenByDescending(pair => pair.Value.LastSeen)
            .First();
        bool changed = false;

        foreach (var candidate in group)
        {
            if (string.Equals(candidate.Key, survivor.Key, StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            MergeRecords(survivor.Value, candidate.Value);
            if (_devices.TryRemove(candidate.Key, out _))
            {
                changed = true;
            }
        }

        return changed;
    }

    private static bool CanMergeDuplicateDevices(Device left, Device right)
    {
        if (string.Equals(left.DeviceId, right.DeviceId, StringComparison.OrdinalIgnoreCase) ||
            left.Platform != right.Platform ||
            !string.Equals(left.Name?.Trim(), right.Name?.Trim(), StringComparison.OrdinalIgnoreCase) ||
            !SameOptionalIdentity(left.Manufacturer, right.Manufacturer) ||
            !SameOptionalIdentity(left.Model, right.Model))
        {
            return false;
        }

        // The same LAN address is the strong signal here: two different phones
        // may have the same model/name, but cannot own the same active address.
        return left.NetworkAddresses.Any(address =>
            right.NetworkAddresses.Any(other =>
                string.Equals(address, other, StringComparison.OrdinalIgnoreCase)));
    }

    private static void MergeRecords(DeviceRecord target, DeviceRecord source)
    {
        bool sourceIsNewer = source.LastSeen >= target.LastSeen;
        target.LastSeen = target.LastSeen >= source.LastSeen
            ? target.LastSeen
            : source.LastSeen;

        foreach (var address in source.Device.NetworkAddresses)
        {
            if (!target.Device.NetworkAddresses.Contains(address, StringComparer.OrdinalIgnoreCase))
            {
                target.Device.NetworkAddresses.Add(address);
            }
        }

        if (source.Device.TrustState == TrustState.Trusted ||
            source.Device.TrustState == TrustState.Blocked)
        {
            target.Device.TrustState = source.Device.TrustState;
        }

        if (sourceIsNewer)
        {
            target.Device.Name = source.Device.Name;
            target.Device.Manufacturer = source.Device.Manufacturer;
            target.Device.Model = source.Device.Model;
            target.Device.AppVersion = source.Device.AppVersion;
            target.Device.ProtocolVersion = source.Device.ProtocolVersion;
            target.Device.SessionPort = source.Device.SessionPort;
            target.Device.DiscoveryPort = source.Device.DiscoveryPort;
            target.Device.Capabilities = source.Device.Capabilities;
        }
    }

    private void NotifyChanged()
    {
        DevicesChanged?.Invoke(this, GetAllDevices());
    }

    private static bool CanCollapseStaleIdentity(
        Device existing,
        DiscoveryMessage incoming,
        string remoteAddress)
    {
        if (existing.ConnectionState != ConnectionState.Disconnected ||
            !string.Equals(existing.Name, DisplayName(
                incoming.Name,
                incoming.Manufacturer,
                incoming.Model),
                StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        var incomingPlatform = Enum.TryParse<DevicePlatform>(
            incoming.Platform,
            true,
            out var parsedPlatform)
            ? parsedPlatform
            : DevicePlatform.Unknown;
        if (existing.Platform != incomingPlatform ||
            !SameOptionalIdentity(existing.Manufacturer, incoming.Manufacturer) ||
            !SameOptionalIdentity(existing.Model, incoming.Model))
        {
            return false;
        }

        return existing.NetworkAddresses.Any(address =>
            string.Equals(address, remoteAddress, StringComparison.OrdinalIgnoreCase));
    }

    private static bool SameOptionalIdentity(string left, string right)
    {
        if (string.IsNullOrWhiteSpace(left) || string.IsNullOrWhiteSpace(right)) return true;
        return string.Equals(left.Trim(), right.Trim(), StringComparison.OrdinalIgnoreCase);
    }

    private static int ValidDiscoveryPort(int port) =>
        port > 0 && port <= 65535 ? port : Constants.DiscoveryUdpPort;

    private static List<string> MergeAddresses(
        IEnumerable<string>? previous,
        string current)
    {
        var addresses = previous?
            .Where(address => !string.IsNullOrWhiteSpace(address))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList() ?? new List<string>();
        if (!addresses.Contains(current, StringComparer.OrdinalIgnoreCase))
        {
            addresses.Add(current);
        }
        return addresses;
    }

    private static string DisplayName(string name, string manufacturer, string model)
    {
        var candidate = name?.Trim() ?? string.Empty;
        if (!string.IsNullOrWhiteSpace(candidate) &&
            !IsPlaceholderName(candidate) &&
            !string.Equals(candidate, model?.Trim(), StringComparison.OrdinalIgnoreCase))
        {
            return candidate;
        }

        var fallback = string.Join(" ", new[] { manufacturer?.Trim(), model?.Trim() }
            .Where(value => !string.IsNullOrWhiteSpace(value)));
        return string.IsNullOrWhiteSpace(fallback) ? "Android 设备" : fallback;
    }

    private static bool IsPlaceholderName(string value)
    {
        return value.Trim().ToLowerInvariant() switch
        {
            "android" or
            "android device" or
            "android 设备" or
            "phone" or
            "mobile" or
            "unknown" or
            "unknown device" or
            "未命名设备" or
            "null" or
            "none" => true,
            _ => false
        };
    }
}

File: android/lib/core/device_registry.dart
import 'dart:async';

import 'device_model.dart';
import 'discovery_message.dart';

class _DeviceRecord {
  final Device device;
  DateTime lastSeen;

  _DeviceRecord({required this.device, required this.lastSeen});
}

class DeviceRegistry {
  final Map<String, _DeviceRecord> _records = {};
  final StreamController<List<Device>> _controller =
      StreamController<List<Device>>.broadcast();

  Stream<List<Device>> get devicesStream => _controller.stream;

  List<Device> get devices => _records.values.map((r) => r.device).toList();

  void upsertDevice(DiscoveryMessage message, String remoteAddress) {
    final now = DateTime.now();
    bool changed = false;

    if (_records.containsKey(message.deviceId)) {
      final record = _records[message.deviceId]!;
      record.lastSeen = now;

      final dev = record.device;
      final addressIndex = dev.networkAddresses.indexOf(remoteAddress);
      if (addressIndex < 0) {
        // Keep address order stable. Phones with Wi-Fi, hotspot, or VPN
        // interfaces can announce from more than one address; moving the
        // latest packet to index 0 caused the UI to rebuild on every packet.
        dev.networkAddresses.add(remoteAddress);
        changed = true;
      }
      if (dev.connectionState == DeviceConnectionState.disconnected) {
        // Device re-appeared
        _records[message.deviceId] = _DeviceRecord(
          device: Device(
            deviceId: dev.deviceId,
            name: message.name,
            manufacturer: message.manufacturer,
            model: message.model,
            platform: _parsePlatform(message.platform),
            appVersion: message.version,
            protocolVersion: message.protocolVersion,
            sessionPort: message.port,
            capabilities: message.capabilities,
            networkAddresses: dev.networkAddresses,
            connectionState: DeviceConnectionState.discovered,
            trustState: dev.trustState,
          ),
          lastSeen: now,
        );
        changed = true;
      } else if (dev.name != message.name ||
          dev.manufacturer != message.manufacturer ||
          dev.model != message.model ||
          dev.sessionPort != message.port) {
        _records[message.deviceId] = _DeviceRecord(
          device: Device(
            deviceId: dev.deviceId,
            name: message.name,
            manufacturer: message.manufacturer,
            model: message.model,
            platform: dev.platform,
            appVersion: message.version,
            protocolVersion: message.protocolVersion,
            sessionPort: message.port,
            capabilities: message.capabilities,
            networkAddresses: dev.networkAddresses,
            connectionState: dev.connectionState,
            trustState: dev.trustState,
          ),
          lastSeen: now,
        );
        changed = true;
      }
    } else {
      final newDevice = Device(
        deviceId: message.deviceId,
        name: message.name,
        manufacturer: message.manufacturer,
        model: message.model,
        platform: _parsePlatform(message.platform),
        appVersion: message.version,
        protocolVersion: message.protocolVersion,
        sessionPort: message.port,
        capabilities: message.capabilities,
        networkAddresses: [remoteAddress],
        connectionState: DeviceConnectionState.discovered,
        trustState: DeviceTrustState.untrusted,
      );
      _records[message.deviceId] = _DeviceRecord(
        device: newDevice,
        lastSeen: now,
      );
      changed = true;
    }

    // A device ID can change after reinstalling or restoring the app. Reconcile
    // every announcement, not only the narrow case where the old record was
    // already offline, so a stale row cannot remain beside the live row.
    changed = _reconcileDuplicateRecords(message.deviceId) || changed;

    if (changed) {
      _notify();
    }
  }

  void pruneOffline(Duration timeout) {
    final cutoff = DateTime.now().subtract(timeout);
    bool changed = false;

    for (final entry in _records.entries) {
      final record = entry.value;
      // A live TCP session is authoritative. UDP discovery is only presence
      // information and may be lost by an access point for a few intervals.
      if (record.device.connectionState == DeviceConnectionState.connected) {
        continue;
      }
      if (record.lastSeen.isBefore(cutoff) &&
          record.device.connectionState != DeviceConnectionState.disconnected) {
        _records[entry.key] = _DeviceRecord(
          device: Device(
            deviceId: record.device.deviceId,
            name: record.device.name,
            manufacturer: record.device.manufacturer,
            model: record.device.model,
            sessionPort: record.device.sessionPort,
            platform: record.device.platform,
            appVersion: record.device.appVersion,
            protocolVersion: record.device.protocolVersion,
            capabilities: record.device.capabilities,
            networkAddresses: record.device.networkAddresses,
            connectionState: DeviceConnectionState.disconnected,
            trustState: record.device.trustState,
          ),
          lastSeen: record.lastSeen,
        );
        changed = true;
      }
    }

    if (changed) {
      _notify();
    }
  }

  void markSessionDisconnected(String deviceId) {
    final record = _records[deviceId];
    if (record == null ||
        record.device.connectionState != DeviceConnectionState.connected) {
      return;
    }
    final stillPresent =
        DateTime.now().difference(record.lastSeen) <=
        const Duration(seconds: 30);
    _records[deviceId] = _DeviceRecord(
      device: record.device.copyWith(
        connectionState: stillPresent
            ? DeviceConnectionState.discovered
            : DeviceConnectionState.disconnected,
      ),
      lastSeen: record.lastSeen,
    );
    _notify();
  }

  bool _reconcileDuplicateRecords(String preferredDeviceId) {
    final preferred = _records[preferredDeviceId];
    if (preferred == null) return false;

    final group = _records.entries
        .where(
          (entry) =>
              entry.key == preferredDeviceId ||
              _canMergeDuplicateDevices(preferred.device, entry.value.device),
        )
        .toList();
    if (group.length <= 1) return false;

    group.sort((left, right) {
      final leftConnected =
          left.value.device.connectionState == DeviceConnectionState.connected;
      final rightConnected =
          right.value.device.connectionState == DeviceConnectionState.connected;
      if (leftConnected != rightConnected) return rightConnected ? 1 : -1;

      final leftPreferred = left.key == preferredDeviceId;
      final rightPreferred = right.key == preferredDeviceId;
      if (leftPreferred != rightPreferred) return rightPreferred ? 1 : -1;

      return right.value.lastSeen.compareTo(left.value.lastSeen);
    });

    final survivor = group.first;
    bool changed = false;
    for (final candidate in group.skip(1)) {
      _records[survivor.key] = _mergeRecords(survivor.value, candidate.value);
      _records.remove(candidate.key);
      changed = true;
    }
    return changed;
  }

  static bool _canMergeDuplicateDevices(Device left, Device right) {
    if (left.deviceId == right.deviceId ||
        left.platform != right.platform ||
        left.name.trim().toLowerCase() != right.name.trim().toLowerCase() ||
        !_sameOptionalIdentity(left.manufacturer, right.manufacturer) ||
        !_sameOptionalIdentity(left.model, right.model)) {
      return false;
    }

    // A shared active LAN address is the strong signal. Two phones can have
    // the same model and user-facing name, but cannot own the same address at
    // the same time.
    return left.networkAddresses.any(
      (address) => right.networkAddresses.any(
        (other) => address.toLowerCase() == other.toLowerCase(),
      ),
    );
  }

  static bool _sameOptionalIdentity(String left, String right) {
    if (left.trim().isEmpty || right.trim().isEmpty) return true;
    return left.trim().toLowerCase() == right.trim().toLowerCase();
  }

  static _DeviceRecord _mergeRecords(
    _DeviceRecord target,
    _DeviceRecord source,
  ) {
    final sourceIsNewer = !source.lastSeen.isBefore(target.lastSeen);
    final addresses = <String>[...target.device.networkAddresses];
    for (final address in source.device.networkAddresses) {
      if (!addresses.any(
        (existing) => existing.toLowerCase() == address.toLowerCase(),
      )) {
        addresses.add(address);
      }
    }

    final trustState =
        source.device.trustState == DeviceTrustState.trusted ||
            source.device.trustState == DeviceTrustState.blocked
        ? source.device.trustState
        : target.device.trustState;
    final mergedDevice = target.device.copyWith(
      name: sourceIsNewer ? source.device.name : target.device.name,
      manufacturer: sourceIsNewer
          ? source.device.manufacturer
          : target.device.manufacturer,
      model: sourceIsNewer ? source.device.model : target.device.model,
      appVersion: sourceIsNewer
          ? source.device.appVersion
          : target.device.appVersion,
      protocolVersion: sourceIsNewer
          ? source.device.protocolVersion
          : target.device.protocolVersion,
      sessionPort: sourceIsNewer
          ? source.device.sessionPort
          : target.device.sessionPort,
      capabilities: sourceIsNewer
          ? source.device.capabilities
          : target.device.capabilities,
      networkAddresses: addresses,
      trustState: trustState,
    );
    return _DeviceRecord(
      device: mergedDevice,
      lastSeen: sourceIsNewer ? source.lastSeen : target.lastSeen,
    );
  }

  static DevicePlatform _parsePlatform(String p) {
    switch (p.toLowerCase()) {
      case 'android':
        return DevicePlatform.android;
      case 'windows':
        return DevicePlatform.windows;
      case 'linux':
        return DevicePlatform.linux;
      case 'macos':
        return DevicePlatform.macos;
      case 'ios':
        return DevicePlatform.ios;
      default:
        return DevicePlatform.unknown;
    }
  }

  void _notify() {
    if (!_controller.isClosed) {
      _controller.add(devices);
    }
  }

  void dispose() {
    _controller.close();
  }
}
